notifications/manager.go

```diff
@@ -0,0 +1,205 @@
+package notifications
+
+import (
+	"context"
+	"time"
+
+	"github.com/lightninglabs/loop/swapserverrpc"
+	"google.golang.org/grpc"
+)
+
+type NotificationType int
+
+const (
+	NotificationTypeUnknown NotificationType = iota
+	NotificationTypeReservation
+)
+
+type NotificationsClient interface {
+	SubscribeNotifications(ctx context.Context,
+		in *swapserverrpc.SubscribeNotificationsRequest,
+		opts ...grpc.CallOption) (
+		swapserverrpc.SwapServer_SubscribeNotificationsClient, error)
+}
+
+// Config contains all the services that the notification manager needs to
+// operate.
+type Config struct {
+	// Client is the client used to communicate with the swap server.
+	Client NotificationsClient
+
+	// FetchL402 is the function used to fetch the l402 token.
+	FetchL402 func(context.Context) error
+}
+
+// Manager is a manager for notifications that the swap server sends to the
+// client.
+type Manager struct {
+	cfg *Config
+
+	hasL402 bool
+
+	subscribers map[NotificationType][]subscriber
+}
+
+// NewManager creates a new notification manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:         cfg,
+		subscribers: make(map[NotificationType][]subscriber),
+	}
+}
+
+type subscriber struct {
+	subCtx   context.Context
+	recvChan interface{}
+}
+
+// SubscribeReservations subscribes to the reservation notifications.
+func (m *Manager) SubscribeReservations(ctx context.Context,
+) <-chan *swapserverrpc.ServerReservationNotification {
+
+	// We'll create a channel that we'll use to send the notifications to the
+	// caller.
+	notifChan := make(
+		chan *swapserverrpc.ServerReservationNotification, //nolint:lll
+	)
+	sub := subscriber{
+		subCtx:   ctx,
+		recvChan: notifChan,
+	}
+
+	m.subscribers[NotificationTypeReservation] = append(
```

Comment:
I think this is a race, because `m.subscribers` is accessed here and also in `handleNotification`.

To avoid this, we can either protect it with a mutex or (more common in Loop codebase) pass new subscriptions to the event loop via a channel.